design/host-discovery.md

```diff
@@ -0,0 +1,292 @@
+cluster and is not misconfigured (PXE booting from the wrong
+environment) or a malicious actor.
+
+#### Replacing a faulty node
+
+*As a data center staff member, I want to replace a faulty node in a
+cluster quickly.*
+
+The idea here is to be able to take a “broken” host out of service and
+replace it with a standby host without having access to the kubernetes
+API or UI at all. If we assume the broken host is broken enough that
+it is completely offline, it should be possible to remove it from the
+rack and replace it with another physical host, which is then
+provisioned automatically.
+
+## Design Details
+
+When Ironic is configured to support its existing discovery feature,
+unknown hosts that boot the Ironic agent will have their hardware
+details registered with `ironic-inspector` and have a node created in
+the Ironic database. Typically this happens via the PXE boot
+configuration, but in the future it could happen in other ways such as
+having a bootable ISO image with the agent on it. How the agent ends
+up on the host should not matter in the rest of the workflow.
+
+The data collected for an unknown host includes the same inspection
+data available for any other host, but does not include the BMC
+credentials. We could use the data to partially create a
+`BareMetalHost` resource for the host by polling the Ironic API to
+compare the list of nodes it knows about with the list of host
+resources. This polling is independent of the normal controller
+reconciliation loop, and should run in a single separate goroutine in
+the `baremetal-operator` process, started after the process wins the
+leader election. A command line flag will control the frequency of
+polling with a default of every `60` seconds. If the frequency is set
+to `0` then polling for discovered hosts will be disabled.
+
+The discovery loop will create a `BareMetalHost` with its name set to
+the Ironic UUID for the discovered node. This will ensure a unique
+name, and will allow the `IronicProvisioner` in the
+`baremetal-operator` to match the host resource to the Ironic node
+automatically. Resource names are immutable, so this may make it
+difficult for a user to keep track of hosts, but other discovered
+identifying information such as the MAC address, IP address(es), and
+serial numbers will also be available. Users can also place data in
+the `spec.description` field, if necessary.
+
+The [state
+machine](https://github.com/metal3-io/baremetal-operator/blob/master/docs/baremetalhost-states.md)
+for the `baremetal-operator` will expand so that hosts in the
+`Discovered` state that appear in the Ironic database can have their
+hardware details fetched from the `ironic-inspector` API, even if the
+host has no BMC credentials. When a new host appears without
+credentials, it will move from `Created` to a new state `Importing`
+during which the details will be fetched from `ironic-inspector`. The
+host will continue to reconcile in the `Importing` state until details
```

Comment:
Aren't the details available immediately? The thing that creates the host in ironic is IPA reporting the inspection results, isn't it? It seems like we could just handle this in the Discovered state without needing to create a separate state.
As written, this would appear to create problems when a user manually creates a Host without adding the credentials initially, which is a thing we will likely also want to support.

Reply:
I've tried to address this in the new draft. If a host is created without credentials (by something other than the new controller, say the user or an installer tool) and then the host is powered on, we will have to wait for the details to show up in the ironic database.